util/licence-checker.py

```diff
@@ -0,0 +1,357 @@
+
+import argparse
+import subprocess
+from pathlib import Path
+from types import SimpleNamespace
+import logging
+import fnmatch
+from itertools import groupby
+
+import hjson
+
+
+class LicenceHeader(object):
+    """Represents the licence header we want to insert"""
+    def __init__(self, text):
+        self._lines = text.strip().splitlines()
+
+    def __getitem__(self, idx):
+        return self._lines.__getitem__(idx)
+
+    def __len__(self):
+        return self._lines.__len__()
+
+    def numbered_lines(self, skip=0):
+        """Returns an iterator of (line_no, line_text).
+
+        `line_no` counts from 1, and is for humans to count line numbers with.
+        use `skip_lines` to skip enumerating the first few lines.
+        """
+        return enumerate(self._lines[skip:], start=1 + skip)
+
+    @property
+    def first_word(self):
+        (first_word, _) = self._lines[0].split(' ', 1)
+        return first_word
+
+
+class LineCommentStyle(object):
+    """Helpers for line-style comments."""
+    def __init__(self, prefix):
+        self.comment_prefix = str(prefix)
+        self.first_line_prefix = self.comment_prefix
+
+    def search_line(self, licence_first_word):
+        return self.comment_prefix + ' ' + licence_first_word
+
+    def expected_full_line(self, licence_line):
+        return self.comment_prefix + ' ' + licence_line
+
+
+class DifferentFirstLineCommentStyle(LineCommentStyle):
+    """Some files have a different allowable prefix for their first line."""
+    def __init__(self, prefix, first_line_prefix):
+        LineCommentStyle.__init__(self, prefix)
+        self.first_line_prefix = first_line_prefix
+
+
+class BlockCommentStyle(object):
+    """Helpers for block-style comments."""
+    def __init__(self, prefix, suffix):
+        self.comment_prefix = str(prefix)
+        self.comment_suffix = str(suffix)
+        self.first_line_prefix = self.comment_prefix
+
+    def search_line(self, licence_first_word):
+        return self.comment_prefix + ' ' + licence_first_word
+
+    def expected_full_line(self, licence_line):
+        return self.comment_prefix + ' ' + licence_line + ' ' + self.comment_suffix
+
+
+SLASH_SLASH = LineCommentStyle("//")
+HASH = LineCommentStyle("#")
+
+# (Priortised) Mapping of file name suffixes to CommentStyle object.
+# If the suffix of your file does not match one of these, it will not be
+# checked.
+COMMENT_CHARS = [
+    # Hardware Files
+    ([".svh", ".sv", ".sv.tpl"], SLASH_SLASH),  # SystemVerilog
+    # Hardware Build Systems
+    ([".tcl"], HASH),  # tcl
+    ([".core", ".core.tpl"], DifferentFirstLineCommentStyle("#", "CAPI=2")),  # FuseSoC Core Files
+    (["Makefile", ".mk"], HASH),  # Makefiles
+    ([".ys"], HASH),  # Yosys script
+    ([".waiver"], HASH),  # Wavier files
+    ([".vlt"], SLASH_SLASH),  # More wavier files
+    ([".vbl"], HASH),  # Verible configuration
+    ([".el", ".el.tpl"], SLASH_SLASH), # Exclusion List
+    ([".do"], SLASH_SLASH),
```

Comment:
These are typically (tool-specific variants) of TCL files, and we already have two different comment styles in the tree for `.do` files.

Reply:
So this should just go under TCL?

Reply:
Well, they use `//` comments at the moment, which isn't standard TCL. Since `.do` files are not standardized in any way and every tool can do different things with them, forcing any type of comment is likely to go wrong.

Reply:
I realise that this would be a more major rework, but another thing you could do is just list extensions that you want to check and then try to infer the comment syntax on reading the file. Something like "Look for first line of header as a substring, then figure out any left/right comment syntax".

Maybe I've missed a good reason to have the explicit comment styles? (For example, do we want to disallow C files with `/* ... */` comments for the licence header?)

Reply:
The license checker is clearly more strict than it needs to be in legal terms, it's more of a "license style checker". So I'm happy with it to enforce a particular style for a file type, and not get into the business of inferring a comment style or something similar.

Reply:
Yeah, I felt I was on shaky ground inferring the first line of where we've found the comment, I would not feel confident in a tool that inferred comment style entirely.